### PR TITLE
8282509: [exploded image] ResolvedClassTest fails with similar output

### PR DESCRIPTION
**Pull request by backwaterred**
This is a tentative solution to a failure observed on AIX. The solution is tentative because I require help to ensure that something deeper and more problematic is not happening.

The test fails because the output produced by PrintCompilation produces `LambdaForm$MH/0x00000007c0002400` instead of `Invokers$Holder` as it does on other platforms. There is one other place the output is different, when `DirectMethodHandle$Holder` is replaced with `LambdaForm$DMH/0x00000007c0001c00`. Ignoring these name changes, the output of PrintCompilation is identical. I observe the same compilations (including the OSR/non-OSR, and same level) in the same order.

I would be grateful for help understanding the root of the difference behind the change. I have a few ideas, but I will let you build your own interpretations free from my potentially incorrect understanding (i.e. no spoilers). Thanks in advance.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8282509](https://bugs.openjdk.java.net/browse/JDK-8282509): [exploded image] ResolvedClassTest fails with similar output

### Reviewers
 * [Martin Doerr](https://openjdk.java.net/census#mdoerr) (@TheRealMDoerr - **Reviewer**)
 * [Dean Long](https://openjdk.java.net/census#dlong) (@dean-long - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7701/head:pull/7701` \
`$ git checkout pull/7701`

Update a local copy of the PR: \
`$ git checkout pull/7701` \
`$ git pull https://git.openjdk.java.net/jdk pull/7701/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7701`

View PR using the GUI difftool: \
`$ git pr show -t 7701`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7701.diff">https://git.openjdk.java.net/jdk/pull/7701.diff</a>

</details>

### CODE REVIEW

**Comment by backwaterred** [comment]
Dean Long mentioned on the issue that they were able to reproduce the issue with an exploded image. I don't have access to JBS (yet!) to comment there directly, so I'll mention here that I am building with make images, and not the default target.

**Comment by backwaterred** [comment]
I believe I have a better handle on this now. In a lucky break, some logging output I added ended up in `build/.../images/jdk/lib/classlist` and caused a build failure on x86. The same failure (with the same logging code) did not occur on AIX. I now believe the difference in test output is due to CDS, which I am surprised to learn is not implemented on [AIX](https://wiki.openjdk.java.net/display/PPCAIXPort).

My working understanding is as follows. For systems which support CDS, PrintCompilation sees the code as coming from Invokers$Holder in Invokers.java where is is shared among JVM instances. On AIX, CDS is not (fully?) implemented, and the code is not stored in Invokers$Holder. Instead, the code is referred to by a LambdaForm at the displayed memory address.

With this piece of the puzzle in place (and hopefully confirmed by others), I believe the check proposed here to be a sane choice to resolve this failure.

**Review by TheRealMDoerr — APPROVED** [review]
Looks reasonable. Please update the PR title: "[exploded image] ResolvedTestClass fails with similar output"

**Comment by backwaterred** [comment]
Thanks for the review @TheRealMDoerr. Should I wait for another review, or is one sufficient in this case?

---

For future reference, it looks to me that this failure is related to missing [pre-generated invokers](https://bugs.openjdk.java.net/browse/JDK-8247536). As I understand it, these are either not created, or just not linked in the exploded image, and are never created on AIX.

**Comment by TheRealMDoerr** [comment]
I've added your additional hint to the issue and linked both issues. Thanks for your investigation.
Your fix could be considered as trivial, but the underlying issue is not. So, please wait for a 2nd review!
The test name in title is not yet correct: Please use "ResolvedClassTest"

**Review by dean-long — APPROVED** [review]
This does seem like the best fix.

**Comment by backwaterred** [comment]
Thanks for the reviews.

/integrate

**Pull request closed without merging**
